Fix type errors in orm.code modules

If the orm code modules are removed from the pre-commit mypy excluded list, below are the current errors.

1. Issues assuming that a computer has been set
2. Issues about what type is used for the executable path (see also https://github.com/aiidateam/aiida-core/issues/5797#issuecomment-1332433139)

It would be good to get these fixed

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

aiida/orm/nodes/data/code/portable.py:57: error: Incompatible types in assignment (expression has type "str", variable has type "PurePath")  [assignment]
aiida/orm/nodes/data/code/portable.py:58: error: Argument 1 to "put_object_from_tree" of "NodeRepository" has incompatible type "Path"; expected "str"  [arg-type]
aiida/orm/nodes/data/code/portable.py:94: error: Incompatible return value type (got "PurePath", expected "Path")  [return-value]
aiida/orm/nodes/data/code/installed.py:45: error: Incompatible types in assignment (expression has type "str", variable has type "PurePath")  [assignment]
aiida/orm/nodes/data/code/installed.py:74: error: Item "None" of "Optional[Computer]" has no attribute "get_transport"  [union-attr]
aiida/orm/nodes/data/code/installed.py:93: error: Item "None" of "Optional[Computer]" has no attribute "pk"  [union-attr]
aiida/orm/nodes/data/code/installed.py:100: error: Incompatible return value type (got "PurePath", expected "Path")  [return-value]
aiida/orm/nodes/data/code/installed.py:111: error: Item "None" of "Optional[Computer]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:144: error: Item "None" of "Optional[Computer]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:152: error: Item "None" of "Optional[Computer]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:160: error: Item "None" of "Optional[Computer]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:169: error: Item "None" of "Optional[Computer]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:208: error: Item "List[Any]" of "Union[List[Any], Any]" has no attribute "pk"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:212: error: Value of type "Optional[List[Any]]" is not indexable  [index]
aiida/orm/nodes/data/code/legacy.py:306: error: Item "List[Any]" of "Union[List[Any], Any]" has no attribute "label"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:308: error: Item "List[Any]" of "Union[List[Any], Any]" has no attribute "pk"  [union-attr]
aiida/orm/nodes/data/code/legacy.py:352: error: Item "None" of "Optional[Computer]" has no attribute "get_transport"  [union-attr]
Found 17 errors in 3 files (checked 728 source files)
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -128,9 +128,6 @@ repos:
                 aiida/orm/nodes/data/array/bands.py|
                 aiida/orm/nodes/data/array/trajectory.py|
                 aiida/orm/nodes/data/cif.py|
-                aiida/orm/nodes/data/code/installed.py|
-                aiida/orm/nodes/data/code/legacy.py|
-                aiida/orm/nodes/data/code/portable.py|
                 aiida/orm/nodes/data/remote/base.py|
                 aiida/orm/nodes/data/structure.py|
                 aiida/orm/nodes/data/upf.py|

--- a/aiida/orm/nodes/data/code/abstract.py
+++ b/aiida/orm/nodes/data/code/abstract.py
@@ -76,7 +76,7 @@ class AbstractCode(Data, metaclass=abc.ABCMeta):
         """
 
     @abc.abstractmethod
-    def get_executable(self) -> pathlib.Path:
+    def get_executable(self) -> pathlib.PurePosixPath:
         """Return the executable that the submission script should execute to run the code.
 
         :return: The executable to be called in the submission script.
@@ -129,7 +129,15 @@ class AbstractCode(Data, metaclass=abc.ABCMeta):
         :return: The full label of the code.
         """
 
-    @Data.label.setter  # type: ignore
+    @property
+    def label(self) -> str:
+        """Return the label.
+
+        :return: The label.
+        """
+        return self.backend_entity.label
+
+    @label.setter
     def label(self, value: str) -> None:
         """Set the label.
 

--- a/aiida/orm/nodes/data/code/containerized.py
+++ b/aiida/orm/nodes/data/code/containerized.py
@@ -36,7 +36,7 @@ class ContainerizedCode(InstalledCode):
         self.image_name = image_name
 
     @property
-    def filepath_executable(self) -> pathlib.PurePath:
+    def filepath_executable(self) -> pathlib.PurePosixPath:
         """Return the filepath of the executable that this code represents.
 
         .. note:: This is overridden from the base class since the path does not have to be absolute.

--- a/aiida/orm/nodes/data/code/installed.py
+++ b/aiida/orm/nodes/data/code/installed.py
@@ -14,6 +14,8 @@ storing the absolute filepath of the relevant executable and the computer on whi
 represented by an instance of :class:`aiida.orm.computers.Computer`. Each time a :class:`aiida.engine.CalcJob` is run
 using an ``InstalledCode``, it will run its executable on the associated computer.
 """
+from __future__ import annotations
+
 import pathlib
 
 import click
@@ -42,7 +44,7 @@ class InstalledCode(Code):
         """
         super().__init__(**kwargs)
         self.computer = computer
-        self.filepath_executable = filepath_executable
+        self.filepath_executable = filepath_executable  # type: ignore[assignment]
 
     def _validate(self):
         """Validate the instance by checking that a computer has been defined.
@@ -92,12 +94,30 @@ class InstalledCode(Code):
         type_check(computer, Computer)
         return computer.pk == self.computer.pk
 
-    def get_executable(self) -> pathlib.Path:
+    def get_executable(self) -> pathlib.PurePosixPath:
         """Return the executable that the submission script should execute to run the code.
 
         :return: The executable to be called in the submission script.
         """
         return self.filepath_executable
+
+    @property  # type: ignore[override]
+    def computer(self) -> Computer:
+        """Return the computer of this code."""
+        assert self.backend_entity.computer is not None
+        return Computer.from_backend_entity(self.backend_entity.computer)
+
+    @computer.setter
+    def computer(self, computer: Computer) -> None:
+        """Set the computer of this code.
+
+        :param computer: A `Computer`.
+        """
+        if self.is_stored:
+            raise exceptions.ModificationNotAllowed('cannot set the computer on a stored node')
+
+        type_check(computer, Computer, allow_none=False)
+        self.backend_entity.computer = computer.backend_entity
 
     @property
     def full_label(self) -> str:
@@ -111,12 +131,12 @@ class InstalledCode(Code):
         return f'{self.label}@{self.computer.label}'
 
     @property
-    def filepath_executable(self) -> pathlib.PurePath:
+    def filepath_executable(self) -> pathlib.PurePosixPath:
         """Return the absolute filepath of the executable that this code represents.
 
         :return: The absolute filepath of the executable.
         """
-        return pathlib.PurePath(self.base.attributes.get(self._KEY_ATTRIBUTE_FILEPATH_EXECUTABLE))
+        return pathlib.PurePosixPath(self.base.attributes.get(self._KEY_ATTRIBUTE_FILEPATH_EXECUTABLE))
 
     @filepath_executable.setter
     def filepath_executable(self, value: str) -> None:
@@ -126,7 +146,7 @@ class InstalledCode(Code):
         """
         type_check(value, str)
 
-        if not pathlib.PurePath(value).is_absolute():
+        if not pathlib.PurePosixPath(value).is_absolute():
             raise ValueError('the `filepath_executable` should be absolute.')
 
         self.base.attributes.set(self._KEY_ATTRIBUTE_FILEPATH_EXECUTABLE, value)

--- a/aiida/orm/nodes/data/code/legacy.py
+++ b/aiida/orm/nodes/data/code/legacy.py
@@ -86,7 +86,7 @@ class Code(AbstractCode):
         type_check(computer, orm.Computer)
         return computer.pk == self.get_remote_computer().pk
 
-    def get_executable(self) -> pathlib.Path:
+    def get_executable(self) -> pathlib.PurePosixPath:
         """Return the executable that the submission script should execute to run the code.
 
         :return: The executable to be called in the submission script.
@@ -96,7 +96,7 @@ class Code(AbstractCode):
         else:
             exec_path = self.get_remote_exec_path()
 
-        return pathlib.Path(exec_path)
+        return pathlib.PurePosixPath(exec_path)
 
     def hide(self):
         """
@@ -140,16 +140,17 @@ class Code(AbstractCode):
                     self.base.repository.put_object_from_filelike(handle, os.path.split(filename)[1])
 
     def __str__(self):
-        local_str = 'Local' if self.is_local() else 'Remote'
-        computer_str = self.computer.label
-        return f"{local_str} code '{self.label}' on {computer_str}, pk: {self.pk}, uuid: {self.uuid}"
+        if self.computer is None:
+            return f"Local code '{self.label}' pk: {self.pk}, uuid: {self.uuid}"
+
+        return f"Remote code '{self.label}' on {self.computer.label} pk: {self.pk}, uuid: {self.uuid}"
 
     def get_computer_label(self):
         """Get label of this code's computer."""
         warn_deprecation(
             'This method is deprecated, use the `InstalledCode.computer.label` property instead.', version=3
         )
-        return 'repository' if self.is_local() else self.computer.label
+        return 'repository' if self.computer is None else self.computer.label
 
     @property
     def full_label(self):
@@ -157,7 +158,7 @@ class Code(AbstractCode):
 
         Returns label of the form <code-label>@<computer-name>.
         """
-        return f'{self.label}@{"repository" if self.is_local() else self.computer.label}'
+        return f'{self.label}@{"repository" if self.computer is None else self.computer.label}'
 
     def relabel(self, new_label):
         """Relabel this code.
@@ -165,10 +166,10 @@ class Code(AbstractCode):
         :param new_label: new code label
         """
         warn_deprecation('This method is deprecated, use the `label` property instead.', version=3)
-        # pylint: disable=unused-argument
-        suffix = f'@{self.computer.label}'
-        if new_label.endswith(suffix):
-            new_label = new_label[:-len(suffix)]
+        if self.computer is not None:
+            suffix = f'@{self.computer.label}'
+            if new_label.endswith(suffix):
+                new_label = new_label[:-len(suffix)]
 
         self.label = new_label
 
@@ -205,11 +206,15 @@ class Code(AbstractCode):
         elif query.count() > 1:
             codes = query.all(flat=True)
             retstr = f"There are multiple codes with label '{label}', having IDs: "
-            retstr += f"{', '.join(sorted([str(c.pk) for c in codes]))}.\n"
+            retstr += f"{', '.join(sorted([str(c.pk) for c in codes]))}.\n"  # type: ignore[union-attr]
             retstr += ('Relabel them (using their ID), or refer to them with their ID.')
             raise MultipleObjectsError(retstr)
         else:
-            return query.first()[0]
+            result = query.first()
+            if not result:
+                raise NotExistent(f"code '{label}' does not exist.")
+
+            return result[0]
 
     @classmethod
     def get(cls, pk=None, label=None, machinename=None):
@@ -303,9 +308,9 @@ class Code(AbstractCode):
         valid_codes = query.all(flat=True)
 
         if labels:
-            return [c.label for c in valid_codes]
+            return [c.label for c in valid_codes]  # type: ignore[union-attr]
 
-        return [c.pk for c in valid_codes]
+        return [c.pk for c in valid_codes]  # type: ignore[union-attr]
 
     def _validate(self):
         super()._validate()
@@ -346,6 +351,9 @@ class Code(AbstractCode):
             version=3
         )
         filepath = self.get_remote_exec_path()
+
+        if self.computer is None:
+            raise exceptions.ValidationError('checking the remote exec path is not available for a local code.')
 
         try:
             with override_log_level():  # Temporarily suppress noisy logging

--- a/aiida/orm/nodes/data/code/portable.py
+++ b/aiida/orm/nodes/data/code/portable.py
@@ -17,6 +17,8 @@ storage. The ``filepath_executable`` should indicate the filename of the executa
 :class:`aiida.engine.CalcJob` is run using a ``PortableCode``, the uploaded files will be automatically copied to the
 working directory on the selected computer and the executable will be run there.
 """
+from __future__ import annotations
+
 import pathlib
 
 import click
@@ -54,8 +56,8 @@ class PortableCode(Code):
         """
         super().__init__(**kwargs)
         type_check(filepath_files, pathlib.Path)
-        self.filepath_executable = filepath_executable
-        self.base.repository.put_object_from_tree(filepath_files)
+        self.filepath_executable = filepath_executable  # type: ignore[assignment]
+        self.base.repository.put_object_from_tree(str(filepath_files))
 
     def _validate(self):
         """Validate the instance by checking that an executable is defined and it is part of the repository files.
@@ -86,7 +88,7 @@ class PortableCode(Code):
         """
         return True
 
-    def get_executable(self) -> pathlib.Path:
+    def get_executable(self) -> pathlib.PurePosixPath:
         """Return the executable that the submission script should execute to run the code.
 
         :return: The executable to be called in the submission script.
@@ -123,12 +125,12 @@ class PortableCode(Code):
         return self.label
 
     @property
-    def filepath_executable(self) -> pathlib.PurePath:
+    def filepath_executable(self) -> pathlib.PurePosixPath:
         """Return the relative filepath of the executable that this code represents.
 
         :return: The relative filepath of the executable.
         """
-        return pathlib.PurePath(self.base.attributes.get(self._KEY_ATTRIBUTE_FILEPATH_EXECUTABLE))
+        return pathlib.PurePosixPath(self.base.attributes.get(self._KEY_ATTRIBUTE_FILEPATH_EXECUTABLE))
 
     @filepath_executable.setter
     def filepath_executable(self, value: str) -> None:
@@ -138,7 +140,7 @@ class PortableCode(Code):
         """
         type_check(value, str)
 
-        if pathlib.PurePath(value).is_absolute():
+        if pathlib.PurePosixPath(value).is_absolute():
             raise ValueError('The `filepath_executable` should not be absolute.')
 
         self.base.attributes.set(self._KEY_ATTRIBUTE_FILEPATH_EXECUTABLE, value)

--- a/tests/orm/data/code/test_abstract.py
+++ b/tests/orm/data/code/test_abstract.py
@@ -23,9 +23,9 @@ class MockCode(AbstractCode):
         """Return whether the code can run on a given computer."""
         return True
 
-    def get_executable(self) -> pathlib.Path:
+    def get_executable(self) -> pathlib.PurePosixPath:
         """Return the executable that the submission script should execute to run the code."""
-        return pathlib.Path('/bin/executable')
+        return pathlib.PurePosixPath('/bin/executable')
 
     @property
     def full_label(self) -> str:

--- a/tests/orm/data/code/test_installed.py
+++ b/tests/orm/data/code/test_installed.py
@@ -43,12 +43,6 @@ def test_validate(aiida_localhost):
     filepath_executable = '/usr/bin/bash'
     code = InstalledCode(computer=aiida_localhost, filepath_executable=filepath_executable)
 
-    code.computer = None
-    assert code.computer is None
-
-    with pytest.raises(ValidationError, match='The `computer` is undefined.'):
-        code.store()
-
     code.computer = aiida_localhost
     code.base.attributes.set(code._KEY_ATTRIBUTE_FILEPATH_EXECUTABLE, None)  # pylint: disable=protected-access
 

